content/testing.md

```diff
@@ -1,23 +1,158 @@
+
+Here is an example of such a value being generated.
+
+```go
+{!generic.md!}
+```
+
+!!! warning
+
+    While randomly generating valid inputs makes sense, randomly generating invalid inputs does not. In the case of invalid inputs, it is much better to have an exhaustive list of all types of cases that are expected to be invalid and always test each one of them.
+
+### Parallelization
+
+Since the version `1.7` of Go, tests can be run in parallel. This can be done by calling [`t.Parallel`](https://pkg.go.dev/testing#T.Parallel) in each subtest. Calling this function signals that the test is to be run in parallel with (and only with) other parallel tests, and the amount of tests running in parallel is limited by the value of [`runtime.GOMAXPROCS`](https://pkg.go.dev/runtime#GOMAXPROCS).
+
+There are multiple advantages to parallelizing tests:
+
+* It ensures that regardless of the order in which inputs are given, components behave as expected.
+* It maximizes performance, which in turns results in a faster CI and a faster workflow for everyone.
+* By making tests run faster, it allows you to write more tests and therefore produce more actionable data to find bugs as well as improve tests cases and coverage.
+* It makes it possible to ensure that the components you test are concurrency-safe
+
+!!! danger "Parallelizing table-driven tests"
+
+    When it comes to table-driven tests, a common caveat developers fall into is to call `t.Parallel` in their subtest without capturing the loop variable with their test case.
+
+    Here is an example of how it should be done:
+
+    ```go
+{!table_driven_parallel.md!}
+    ```
+
+### Standard `testing` Package
+
+The standard [`testing`](https://pkg.go.dev/testing) package is very powerful, and does not require additional frameworks to be used efficiently. The only exception we make to that are the `stretchr/testify/assert` and `stretchr/testify/require` packages which we use only for convenience, as they expose assertion functions that produce consistent outputs and make tests easy to understand.
+
+#### Subtests and Sub-benchmarks
+
+The `testing` package exposes a `Run` method on the `T` type which makes it possible to nest tests within tests. This can be very useful, as it enables creating a hierarchical structure within a test.
+
+??? example "index_internal_test.go"
+
+    ```go
+{!index_internal_test.md!}
+    ```
+
+#### Table-Driven Tests
+
+It makes a lot of sense to use subtests when testing the behavior of complex components, but it is better to use [table-driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) when testing a simple function with a given output for a given input, and that there are many cases to cover. For such cases, table-driven tests massively improve clarity and readability.
```

Comment:
```suggestion
It makes a lot of sense to use subtests when testing the behavior of complex components, but it is better to use [table-driven tests](https://dave.cheney.net/2019/05/07/prefer-table-driven-tests) when testing a simple function with a given output for a given input, and there are many cases to cover. For such cases, table-driven tests massively improve clarity and readability.
```

Also, dubious about "testing a simple function with a given output for a given input". Perhaps it's better to have it as an "expected" output with a given input.